python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,378 @@
+            raise ValueError(
+                f"Received invalid num_to_keep: "
+                f"{self.num_to_keep}. "
+                f"Must be None or non-negative integer."
+            )
+        if self.checkpoint_score_order not in (MAX, MIN):
+            raise ValueError(
+                f"checkpoint_score_order must be either " f'"{MAX}" or "{MIN}".'
+            )
+
+
+class CheckpointManager:
+    """Common checkpoint management and bookkeeping class for Ray Train and Tune.
+
+    This class acts as the common core for checkpoint bookkeeping in Ray ML libraries.
+    On a high level, this manager keeps a reference to all stored checkpoints
+    (both in-memory and on-disk checkpoints). For on-disk checkpoints, it
+    keeps a configured number of checkpoints according to specified metrics.
+
+    The manager supports lazy data writing by utilizing the
+    ``TrackedCheckpoint.commit()`` API, which is only invoked if the checkpoint
+    should be persisted to disk.
+    """
+
+    def __init__(
+        self,
+        checkpoint_strategy: CheckpointStrategy,
+        latest_checkpoint_id: int = 0,
+        delete_fn: Optional[Callable[["TrackedCheckpoint"], None]] = None,
+    ):
+        self._checkpoint_strategy = checkpoint_strategy or CheckpointStrategy()
+
+        # Incremental unique checkpoint ID of this run.
+        self._latest_checkpoint_id = latest_checkpoint_id
+
+        # Used for keeping top K checkpoints.
+        self._top_persisted_checkpoints: List[_HeapCheckpointWrapper] = []
+
+        # Best checkpoint altogether.
+        # Used for exposing best_checkpoint_path.
+        self._best_persisted_checkpoint: Optional[TrackedCheckpoint] = None
+        self._latest_persisted_checkpoint: Optional[TrackedCheckpoint] = None
+        self._latest_memory_checkpoint: Optional[TrackedCheckpoint] = None
+
+        # Checkpoints that are not immediately removed
+        self._checkpoints_to_clean_up = set()
+        self._delete_fn = delete_fn
+
+    def set_delete_fn(self, delete_fn: Optional[Callable[["TrackedCheckpoint"], None]]):
+        """Update the function called to delete persisted checkpoints.
+
+        Args:
+            delete_fn: Function that takes a tracked checkpoint as an argument and
+                deletes it from disk.
+        """
+        self._delete_fn = delete_fn
+
+    def register_checkpoint(self, checkpoint: TrackedCheckpoint):
+        """Register new checkpoint and add to bookkeeping.
+
+        This method will register a new checkpoint and add it to the internal
+        bookkeeping logic. This means the checkpoint manager will decide if
+        this checkpoint should be kept, and if older or worse performing
+        checkpoints should be deleted.
+
+        Subclasses have to implement this method.
+
+        Args:
+            checkpoint: Tracked checkpoint object to add to bookkeeping.
+        """
+        raise NotImplementedError
```

Comment:
Is there some base common logic that we can have here for registering a new checkpoint?

It seems like the general logic can just be this right?

```python
self._decide_what_to_do_with_checkpoint(checkpoint)
```

Reply:
In the current iteration, this is an abstract base class and there is no way to use the manager in a sensible way without customizing the entry point. I've marked it as an ABC to make this clear. We may be able to remove this in the future if we get rid of custom extensions, but until then we should make this explicit.

Reply:
After revisiting this, I've decided we can actually use a common entry point here. Updated the PR.